File: frontend/services/notifications.ts
import axios from "axios";

import { api } from "./api";
import type {
    Notification,
    NotificationListResponse,
    NotificationReadAllResponse,
} from "@/types/notification";


const NOTIFICATION_ERROR = (
    "Notifications are temporarily unavailable."
);


export async function getNotifications(
    limit = 20
): Promise<NotificationListResponse> {

    try {
        const response = await api.get<NotificationListResponse>(
            "/notifications",
            {
                params: { limit },
            }
        );

        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error)) {
            throw new Error(NOTIFICATION_ERROR);
        }

        throw new Error(NOTIFICATION_ERROR);
    }
}


export async function markNotificationRead(
    notificationId: number
): Promise<Notification> {

    try {
        const response = await api.patch<Notification>(
            `/notifications/${notificationId}/read`
        );

        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error)) {
            throw new Error(NOTIFICATION_ERROR);
        }

        throw new Error(NOTIFICATION_ERROR);
    }
}


export async function markAllNotificationsRead(): Promise<
    NotificationReadAllResponse
> {

    try {
        const response = await api.patch<NotificationReadAllResponse>(
            "/notifications/read-all"
        );

        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error)) {
            throw new Error(NOTIFICATION_ERROR);
        }

        throw new Error(NOTIFICATION_ERROR);
    }
}

File: backend/app/api/notifications.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.database.database import get_db
from app.database.models import User
from app.models.notification_model import (
    NotificationListResponse,
    NotificationReadAllResponse,
    NotificationResponse
)
from app.services import notification_service


router = APIRouter(
    prefix="/notifications",
    tags=["Notifications"]
)


@router.get(
    "",
    response_model=NotificationListResponse,
    summary="Get current user's notifications"
)
def get_notifications(
    limit: int = Query(default=20, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):

    return notification_service.list_notifications(
        db,
        current_user.id,
        limit
    )


@router.patch(
    "/read-all",
    response_model=NotificationReadAllResponse,
    summary="Mark all current user's notifications as read"
)
def mark_all_notifications_read(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):

    return notification_service.mark_all_notifications_read(
        db,
        current_user.id
    )


@router.patch(
    "/{notification_id:int}/read",
    response_model=NotificationResponse,
    summary="Mark one current user's notification as read"
)
def mark_notification_read(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):

    return notification_service.mark_notification_read(
        db,
        notification_id,
        current_user.id
    )
